webapp/src/app.tsx

```diff
@@ -1715,16 +1717,20 @@ export class ProjectView
 
     startSimulator(debug?: boolean, clickTrigger?: boolean) {
         pxt.tickEvent('simulator.start');
-        if (!this.shouldStartSimulator()) {
-            pxt.log("Tried to start simulator when we shouldn't");
+        if (!this.shouldStartSimulator() || (this.runToken && this.runToken.isRunning())) {
+            pxt.log("Ignoring call to start simulator, either already running or we shouldn't start.");
             return;
         }
+        if (this.runToken) this.runToken.cancel()
+        this.runToken = new pxt.Util.CancellationToken()
+        this.runToken.startOperation();
         this.saveFileAsync()
-            .then(() => this.runSimulator(debug ? { debug: true } : {}, clickTrigger));
+            .then(() => this.runSimulator({ debug, clickTrigger }, this.runToken));
```

Comment:
remove token from signature, it is handled internally

Reply:
I guess the issue I have with using runToken by default for runSimulator is that we call runSimulator from a number of places and not all of them go through the path of: 
startSimulator -> runSimulator -> stopSimulator.

I'll take another look and see if I can standardize it.